code/modules/projectiles/projectile/special/gravity.dm

```diff
@@ -1,90 +1,73 @@
-/obj/item/projectile/gravityrepulse
-	name = "repulsion bolt"
+/obj/item/projectile/gravity
+	name = "gravity bolt"
 	icon = 'icons/effects/effects.dmi'
 	icon_state = "chronofield"
 	hitsound = 'sound/weapons/wave.ogg'
 	damage = 0
 	damage_type = BRUTE
-	nodamage = 1
-	color = "#33CCFF"
-	var/turf/T
+	nodamage = TRUE
 	var/power = 4
-	var/list/thrown_items = list()
+	var/list/thrown = list()		//normally we wouldn't need this but incase on_hit is called multiple times, yeah.. this is a good idea.
 
-/obj/item/projectile/gravityrepulse/Initialize()
+/obj/item/projectile/gravity/Initialize(mapload)
 	. = ..()
-	var/obj/item/ammo_casing/energy/gravity/repulse/C = loc
-	if(istype(C)) //Hard-coded maximum power so servers can't be crashed by trying to throw the entire Z level's items
-		power = min(C.gun.power, 15)
+	var/obj/item/ammo_casing/energy/gravity/G = loc
+	if(istype(G))
+		power = min(G.gun.power, 15)
 
-/obj/item/projectile/gravityrepulse/on_hit()
+/obj/item/projectile/gravity/on_hit()
 	. = ..()
-	T = get_turf(src)
-	for(var/atom/movable/A in range(T, power))
-		if(A == src || (firer && A == src.firer) || A.anchored || thrown_items[A])
-			continue
-		var/throwtarget = get_edge_target_turf(src, get_dir(src, get_step_away(A, src)))
-		A.throw_at(throwtarget,power+1,1)
-		thrown_items[A] = A
+	var/turf/T = get_turf(src)
+	var/list/tothrow
+	var/list/mobs = list()
+	var/list/objs = list()
+	var/list/cachedrange = range(T, power)
+	for(var/mob/M in cachedrange)
+		mobs += M
+	for(var/obj/O in cachedrange)
+		objs += O
+	tothrow = mobs + objs		//mobs priority
+	var/safety = 50
+	for(var/i in tothrow)
+		if(!safety)
+			break
+		var/atom/movable/AM = i
+		if(can_throw(AM, thrown) && do_the_throw(T, AM, thrown))
+			safety--
 	for(var/turf/F in range(T,power))
 		new /obj/effect/temp_visual/gravpush(F)
 
-/obj/item/projectile/gravityattract
+/obj/item/projectile/gravity/proc/can_throw(atom/movable/AM, list/thrown)
+	return (AM != src) && (AM != firer) && !AM.anchored && !thrown[AM]
+
+/obj/item/projectile/gravity/proc/do_the_throw(turf/origin, atom/movable/AM, list/thrown)
```

Comment:
You better merge these two procs together and make them work with lists to avoid unwarranted proc overhead.